semantic-ast/src/AST/GenerateSyntax.hs

```diff
@@ -146,41 +154,51 @@ debugPrefix (name, Named)     = name
 debugPrefix (name, Anonymous) = "_" <> name
 
 -- | Build Q Constructor for product types (nodes with fields)
-ctorForProductType :: String -> Name -> Maybe Children -> [(String, Field)] -> Q Con
-ctorForProductType constructorName typeParameterName children fields = ctorForTypes constructorName lists where
+ctorForProductType :: DatatypeName -> Maybe Children -> [(String, Field)] -> Q Con
+ctorForProductType constructorName children fields = ctorForTypes constructorName lists where
   lists = annotation : fieldList <> childList
-  annotation = ("ann", varT typeParameterName)
-  fieldList = map (fmap toType) fields
+  annotation = ("ann", varT annParameterName)
+  fieldList = map (fmap (toType)) fields
   childList = toList $ fmap toTypeChild children
+
+  inject t = varT shapeParameterName `appT` t
+
+  toType :: Field -> TypeQ
   toType (MkField required fieldTypes mult) =
-    let ftypes = fieldTypesToNestedSum fieldTypes `appT` varT typeParameterName
+    let ftypes = inject (fieldTypesToNestedSum shapeParameterName fieldTypes `appT` varT annParameterName)
```

Comment:
ensure that the children of each record types are wrapped in an `f` (except for the ann field)